### DISCUSSION TITLE
call to licsenses/ API  endpoint returns correct data but fails to update attributes

### DISCUSSION THREAD

**Discussion by fmg-pete** [keygen-sh/community#144] [General]
Hello!

I am making  a call to the licences/ REST API endpoint, e.g.

https://api.keygen.sh/v1/accounts/<account>/licenses/<id>

I am POSTing this payload:

data: {
     attributes: {
           expiry: "2025-04-26T14:30:00.000Z"
     }
}

The response is the license JSON for the value specified by <id>, which is what I'd expect if no errors occurred.

When I check the license expiry date though, it still shows the old value.

Do I need to end the value in seconds instead of iso8601? Or some maybe other format?

Any clues what I might be missing? 

Thanks in advance. :)

**Comment by ezekg** [comment]
Can you please give me a request ID to look at (the `X-Request-Id` header in the response), or provide the entire request and response bodies? That request should be rejected, because it's missing the `data.type` property, as well as the `data.relationships.policy` relationship.

Also, what do you mean by "it still shows the old value"? You're creating a new license with this `POST` request, not updating an existing license. If you want to update an existing license, you need to `PATCH` the existing license, not `POST` (which creates a new license).

**Reply by fmg-pete** [reply]
Sorry, the formatting here broke my post. The endpoint had the account id and the license id in it, and it appeared from the docs that all I needed to to to update and expiry date in the payload was to reference data.attributes.expiry. If there are other compulsory fields I need for this to work I'm going to have a hard time guessing them. 

I am using the information here, specifically the section "Update a license". 

https://keygen.sh/docs/api/licenses/#licenses-update

The example request shown at the top is exactly what I am sending, what I missed was the PATCH  request type.

This is the full response I get:

```javascript
{
    "id": "345e0ff2-f955-43d8-80e9-7f157be6469b",
    "type": "licenses",
    "attributes": {
        "name": "Pete",
        "key": "880629-888BA5-0F91C9-813EF1-77CAF9-V3",
        "expiry": "2024-03-28T09:10:12.114Z",
        "status": "EXPIRED",
        "uses": 0,
        "suspended": false,
        "scheme": null,
        "encrypted": false,
        "strict": true,
        "floating": true,
        "protected": true,
        "version": null,
        "maxMachines": 3,
        "maxProcesses": null,
        "maxUsers": null,
        "maxCores": null,
        "maxUses": null,
        "requireHeartbeat": true,
        "requireCheckIn": false,
        "lastValidated": "2024-04-26T06:23:11.810Z",
        "lastCheckIn": null,
        "nextCheckIn": null,
        "lastCheckOut": null,
        "metadata": {},
        "created": "2023-03-27T04:29:32.812Z",
        "updated": "2024-04-26T06:23:11.831Z"
    },
    "relationships": {
        "account": {
            "links": {
                "related": "/v1/accounts/bd8403b5-0095-4a49-8a34-594f8c9402f0"
            },
            "data": {
                "type": "accounts",
                "id": "bd8403b5-0095-4a49-8a34-594f8c9402f0"
            }
        },
        "environment": {
            "links": {
                "related": null
            },
            "data": null
        },
        "product": {
            "links": {
                "related": "/v1/accounts/bd8403b5-0095-4a49-8a34-594f8c9402f0/licenses/345e0ff2-f955-43d8-80e9-7f157be6469b/product"
            },
            "data": {
                "type": "products",
                "id": "908b7761-c532-4acc-b284-9ba106b3f250"
            }
        },
        "policy": {
            "links": {
                "related": "/v1/accounts/bd8403b5-0095-4a49-8a34-594f8c9402f0/licenses/345e0ff2-f955-43d8-80e9-7f157be6469b/policy"
            },
            "data": {
                "type": "policies",
                "id": "07949e1c-0ff1-41b0-af60-1033b51c58ca"
            }
        },
        "group": {
            "links": {
                "related": "/v1/accounts/bd8403b5-0095-4a49-8a34-594f8c9402f0/licenses/345e0ff2-f955-43d8-80e9-7f157be6469b/group"
            },
            "data": null
        },
        "users": {
            "links": {
                "related": "/v1/accounts/bd8403b5-0095-4a49-8a34-594f8c9402f0/licenses/345e0ff2-f955-43d8-80e9-7f157be6469b/users"
            },
            "meta": {
                "count": 1
            }
        },
        "machines": {
            "links": {
                "related": "/v1/accounts/bd8403b5-0095-4a49-8a34-594f8c9402f0/licenses/345e0ff2-f955-43d8-80e9-7f157be6469b/machines"
            },
            "meta": {
                "cores": 0,
                "count": 0
            }
        },
        "tokens": {
            "links": {
                "related": "/v1/accounts/bd8403b5-0095-4a49-8a34-594f8c9402f0/licenses/345e0ff2-f955-43d8-80e9-7f157be6469b/tokens"
            }
        },
        "entitlements": {
            "links": {
                "related": "/v1/accounts/bd8403b5-0095-4a49-8a34-594f8c9402f0/licenses/345e0ff2-f955-43d8-80e9-7f157be6469b/entitlements"
            }
        },
        "user": {
            "links": {
                "related": "/v1/accounts/bd8403b5-0095-4a49-8a34-594f8c9402f0/licenses/345e0ff2-f955-43d8-80e9-7f157be6469b/user"
            },
            "data": {
                "type": "users",
                "id": "5e204467-3c68-40c3-85c6-432f1941d404"
            }
        }
    },
    "links": {
        "self": "/v1/accounts/bd8403b5-0095-4a49-8a34-594f8c9402f0/licenses/345e0ff2-f955-43d8-80e9-7f157be6469b"
    }
}
```

**Reply by fmg-pete** [reply]
UPDATE: OK, got it. Corrected payload looks like this:

` {
        "data": {
            "type": "licenses",
            "attributes" : {
                "expiry": isoDate
            }                               
        }`

... and I had to tweak the php_curl request:

`curl_setopt($ch, CURLOPT_HTTPHEADER, array(
        'Accept: */*',
        'X-HTTP-Method-Override: PATCH',
        'Accept: application/vnd.api+json',
        'Connection: Keep-Alive',  
        'Content-Type: application/vnd.api+json',  
        'Authorization: Bearer '.$token
    )); `
